graylog2-server/src/main/java/org/graylog2/system/processing/DBProcessingStatusService.java

```diff
@@ -195,4 +140,51 @@ ProcessingStatusDto save(ProcessingStatusRecorder processingStatusRecorder, Date
                 true, // We want to return the updated document to the caller
                 true));
     }
+
+    public ProcessingNodesState calculateProcessingState(TimeRange timeRange) {
+        final DateTime updateThresholdTimestamp = clock.nowUTC().minus(updateThreshold.toMilliseconds());
+
+        try (DBCursor<ProcessingStatusDto> statusCursor = db.find(activeNodes(updateThresholdTimestamp))) {
+            if (!statusCursor.hasNext()) return ProcessingNodesState.NO_ACTIVE;
+            int activeNodes = 0;
+            int behindNodes = 0;
+            while (statusCursor.hasNext()) {
+                activeNodes++;
+                ProcessingStatusDto nodeProcessingStatus = statusCursor.next();
+                DateTime postIndexing = nodeProcessingStatus.receiveTimes().postIndexing();
+                if (postIndexingBehind(timeRange.getTo(), postIndexing)) {
+                    if (isBusy(nodeProcessingStatus)) {
+                        return ProcessingNodesState.SOME_BUSY;
+                    }
+                    if (postIndexing.isBefore(timeRange.getFrom())) {
+                        behindNodes++;
+                    }
+                }
+            }
+            if (activeNodes == behindNodes) {
+                return ProcessingNodesState.NO_MESSAGES;
+            }
+        }
+
+        return ProcessingNodesState.ALL_UP_TO_DATE;
+    }
+
+    private boolean isBusy(ProcessingStatusDto nodeProcessingStatus) {
+        return nodeProcessingStatus.inputJournal().uncommittedEntries() > 0L;
+    }
+
+    private boolean postIndexingBehind(DateTime timeRangeTo, DateTime postIndexing) {
```

Comment:
I would just inline this method. I found it easier to understand if I see that `.isAfter` call directly